src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -0,0 +1,308 @@
+#include "cFE_util.h"
+#include "osfilesys.h"
+
+struct FS fs[MAX_NUM_FS];
+struct fsobj files[MAX_NUM_FILES];
+
+/*
+ * Notes on this version:
+ * This is not the final implementation of a file system for the cFE on COS
+ * For now the file system is read only and uses static memory for fsobj and FS
+ * The next step is to write a dumb allocator/deallocator and use it for 
+ * new fs, fsobj, and file data
+ * I have been intentional to restrict memory specific details to this file
+ * osfilesys.h and osfiles.c 'should' not care about how FS, fsobj, and data
+ * are stored in memory.
+ */
+
+/******************************************************************************
+** Tar Level Methods
+******************************************************************************/
+
+/* 
+ * Loads the position in memory of linked
+ * Checks for badly linked or no linked tar file.
+ */
+uint32 tar_load()
+{
+    if(! &_binary_cFEfs_tar_size)                         return OS_FS_ERR_DRIVE_NOT_CREATED;
+    if(!_binary_cFEfs_tar_start)                          return OS_FS_ERR_DRIVE_NOT_CREATED;
+    if(&_binary_cFEfs_tar_end < &_binary_cFEfs_tar_start) return OS_FS_ERR_DRIVE_NOT_CREATED;
+    size    = (uint32) &_binary_cFEfs_tar_size;
+    start   = &_binary_cFEfs_tar_start;
+    end     = &_binary_cFEfs_tar_end;
+    return OS_FS_SUCCESS;
+}
+
+/*
+ * parses a loaded tar into fs[0]
+ */
+uint32 tar_parse()
+{
+    if(!start || !end)              return OS_FS_ERR_DRIVE_NOT_CREATED;
+    if(end-start-size)              return OS_FS_ERR_DRIVE_NOT_CREATED;
+    if(!fs[0].devname)              return OS_FS_ERR_DRIVE_NOT_CREATED;
+
+    assert(fs[0].root->filedes==1);
+    struct fsobj *o=fs[0].root;
+    uint32 offset = 0;
+
+    while(offset+start<end){
+        //tar ends after two empty records
+        if( !(offset+start)[0] && !(offset+start)[512]) return OS_SUCCESS;
+
+        if(tar_cphdr(offset,o))     return OS_FS_ERR_DRIVE_NOT_CREATED;
+        if(file_insert(o,&fs[0]))   return OS_FS_ERR_DRIVE_NOT_CREATED;
+        offset += round_up(o->size+500);
+        if(newfile_get(&o))         return OS_FS_ERR_DRIVE_NOT_CREATED;
+    }
```

Comment:
It can be a for loop but it does not end up being as clean.  The size that the offset is incremented by each time is dependant on the size of the data and rounding it up to a 512 byte block.  To me it seems easier to read and immediately understand if that side effect is within the loop.